### PR TITLE
Rocks db erasure decoding

### PR DESCRIPTION
**Pull request by carllin**
#### Problem
Erasure decoding was relying on the old windows instead of RocksDb.

#### Summary of Changes
Update decoding of windows to use new RocksDb style ledger. Needs some elements of 
https://github.com/solana-labs/solana/pull/1888 to build, but tests should be passing after that.

Fixes #

### CODE REVIEW

File: src/erasure.rs [hunk 1]
```diff
@@ -347,101 +348,47 @@ pub fn generate_coding(
...
 // Recover a missing block into window
 //   missing blocks should be None or old...
 //   If not enough coding or data blocks are present to restore
-//    any of the blocks, the block is skipped.
+//   any of the blocks, the block is skipped.
 //   Side effect: old blobs in a block are None'd
-pub fn recover(id: &Pubkey, window: &mut [WindowSlot], start_idx: u64, start: usize) -> Result<()> {
-    let block_start = start - (start % NUM_DATA);
+pub fn recover(
+    id: &Pubkey,
```
Comment (rob-solana) [inline src/erasure.rs]: this can be removed by now, I think

File: src/erasure.rs [hunk 2]
```diff
@@ -587,14 +544,16 @@ pub fn recover(id: &Pubkey, window: &mut [WindowSlot], start_idx: u64, start: us
             locks[n].data()[0]
         );
     }
-    assert!(!corrupt, " {} ", id);
 
-    Ok(())
+    assert!(!corrupt, " {} ", id);
```
Comment (rob-solana) [inline src/erasure.rs]: we have an opportunity to do something about this, now.  if we end up with a corrupt blob, let's not put it in the ledger/window?

File: src/erasure.rs [hunk 3]
```diff
@@ -536,7 +538,10 @@ pub fn recover(
         );
     }
 
-    assert!(!corrupt);
+    if corrupt {
```
Comment (rob-solana) [inline src/erasure.rs]: cool, but we want to remove or mark as bogus the coding blobs from the ledger?  maybe we need a way to say "this is a bogus blob" in the ledger/window...

File: src/erasure.rs [hunk 4]
```diff
@@ -347,101 +348,44 @@ pub fn generate_coding(
...
-//    any of the blocks, the block is skipped.
+//   any of the blocks, the block is skipped.
 //   Side effect: old blobs in a block are None'd
-pub fn recover(id: &Pubkey, window: &mut [WindowSlot], start_idx: u64, start: usize) -> Result<()> {
-    let block_start = start - (start % NUM_DATA);
+pub fn recover(
+    db_ledger: &mut DbLedger,
+    slot: u64,
+    start_idx: u64,
+) -> Result<(Vec<SharedBlob>, Vec<SharedBlob>)> {
```
Comment (sakridge) [inline src/erasure.rs]: What is the result?

Comment (sakridge) [reply src/erasure.rs]: also, the function comment doesn't seem accurate, it doesn't take a window anymore.

File: src/erasure.rs [hunk 5]
```diff
@@ -450,106 +394,115 @@ pub fn recover(id: &Pubkey, window: &mut [WindowSlot], start_idx: u64, start: us
...
-    let size = size.unwrap();
-    for i in block_start..block_end {
-        let j = i % window.len();
-
-        if let Some(b) = &window[j].data {
-            let mut b_wl = b.write().unwrap();
-            for i in b_wl.meta.size..size {
-                b_wl.data[i] = 0;
+    // Add the coding blobs we have into the recovery vector, mark the missing ones
+    for i in coding_start_idx..block_end_idx {
```
Comment (sakridge) [inline src/erasure.rs]: Seems like a lot of logic to duplicate, can be moved into a function?

File: src/erasure.rs (hunk 3 above)
Comment (rob-solana) [reply src/erasure.rs]: yeap.

File: src/erasure.rs (hunk 4 above)
Comment (carllin) [reply src/erasure.rs]: The result is a list of reconstructed data and coding blobs returned to the caller for processing.

Comment (sakridge) [reply src/erasure.rs]: Ok, yea just could be indicated with a comment or something, I think a reader of the the function wouldn't realize it just reading the signature.

Comment (carllin) [reply src/erasure.rs]: For sure, updated the comment

File: src/erasure.rs [hunk 6]
```diff
@@ -2,6 +2,7 @@
 use db_ledger::DbLedger;
 use db_window::{find_missing_coding_indexes, find_missing_data_indexes};
 use packet::{Blob, SharedBlob, BLOB_DATA_SIZE, BLOB_HEADER_SIZE};
+use result::Result as SolanaResult;
```
Comment (rob-solana) [inline src/erasure.rs]: what happened here?

File: src/erasure.rs [hunk 7]
```diff
@@ -545,6 +519,40 @@ pub fn recover(
...
+    match get_blob_result {
+        Err(err) => {
+            error!(
+                "error getting blob at slot: {}, index: {}, error: {:?}",
+                slot, index, err
+            );
+            return Err(ErasureError::InvalidBlobData);
+        }
+        Ok(Some(b)) => {
+            if b.len() <= BLOB_HEADER_SIZE {
```
Comment (rob-solana) [inline src/erasure.rs]: I guess we don't support zero-length blobs...

File: src/erasure.rs [hunk 8]
```diff
@@ -545,6 +519,40 @@ pub fn recover(
...
+                return Err(ErasureError::InvalidBlobData);
+            }
+            blobs.push(Arc::new(RwLock::new(Blob::new(&b))));
+        }
+        Ok(None) => {
+            // Mark the missing memory
+            erasures.push(erasure_index);
+            blobs.push(SharedBlob::default());
+            missing.push(blobs.last_mut().unwrap().clone());
```
Comment (rob-solana) [inline src/erasure.rs]: is this more or less code, compiled, than: 
```
let b = Default::default(); blobs.push(b.clone()); missing.push(b);
``` ?

File: src/erasure.rs (hunk 6 above)
Comment (carllin) [reply src/erasure.rs]: The original erasure module used its own version of Result, so I had to import our more generic result under a different name (SolanaResult) in an earlier commit in order to work with it. I ultimately replaced the erasure-specific Result type with the one in result.rs, so this import from the earlier commit can now be deleted.

File: src/packet.rs [hunk 9]
```diff
@@ -274,6 +275,16 @@ pub const BLOB_FLAG_IS_CODING: u32 = 0x1;
 pub const BLOB_HEADER_SIZE: usize = align!(BLOB_SIZE_END, 64);
 
 impl Blob {
+    pub fn new(data: &[u8]) -> Self {
+        let mut blob = Self::default();
+        let data_len = cmp::min(data.len(), blob.data.len());
```
Comment (rob-solana) [inline src/packet.rs]: this won't always be zero?

Comment (carllin) [reply src/packet.rs]: blob.data is a fixed size array of size 65408, so it shouldn't be

Comment (rob-solana) [reply src/packet.rs]: ok, I see.  we're gonna assert!() that data is small enough to fit?

Comment (carllin) [reply src/packet.rs]: Added a check where new() is called in erasure to make sure the data fits.

**Pull request merged**